sql-to-dbsp-compiler/lib/sqllib/src/lib.rs

```diff
@@ -18,6 +19,291 @@ use std::ops::Add;
 #[derive(Clone)]
 pub struct DefaultOptSemigroup<T>(PhantomData<T>);
 
+// Macro to create variants of a function with 1 argument
+// If there exists a function is f_(x: T) -> S, this creates a function
+// fN(x: Option<T>) -> Option<S>, defined as
+// fN(x) { let x = x?; Some(f_(x)) }.
+#[macro_export]
+macro_rules! some_function1 {
+    ($func_name:ident, $arg_type:ty, $ret_type:ty) => {
+        ::paste::paste! {
+            pub fn [<$func_name N>]( arg: Option<$arg_type> ) -> Option<$ret_type> {
+                let arg = arg?;
+                Some([<$func_name _>](arg))
+            }
+        }
+    }
+}
+
+// Macro to create variants of a function with 2 arguments
+// If there exists a function is f__(x: T, y: S) -> U, this creates
+// three functions:
+// - f_N(x: T, y: Option<S>) -> Option<U>
+// - fN_(x: Option<T>, y: S) -> Option<U>
+// - fNN(x: Option<T>, y: Option<S>) -> Option<U>
+// The resulting functions return Some only if all arguments are 'Some'.
+#[macro_export]
+macro_rules! some_function2 {
+    ($func_name:ident, $arg_type0:ty, $arg_type1:ty, $ret_type:ty) => {
+        ::paste::paste! {
+            pub fn [<$func_name NN>]( arg0: Option<$arg_type0>, arg1: Option<$arg_type1> ) -> Option<$ret_type> {
+                let arg0 = arg0?;
+                let arg1 = arg1?;
+                Some([<$func_name __>](arg0, arg1))
+            }
+
+            pub fn [<$func_name _N>]( arg0: $arg_type0, arg1: Option<$arg_type1> ) -> Option<$ret_type> {
+                let arg1 = arg1?;
+                Some([<$func_name __>](arg0, arg1))
+            }
+
+            pub fn [<$func_name N_>]( arg0: Option<$arg_type0>, arg1: $arg_type1 ) -> Option<$ret_type> {
+                let arg0 = arg0?;
+                Some([<$func_name __>](arg0, arg1))
+            }
+        }
+    }
+}
+
+// Macro to create variants of a function with 4 arguments
+// If there exists a function is f____(x: T, y: S, z: V, w: W) -> U, this creates
+// fifteen functions:
+// - f___N(x: T, y: S, z: V, w: Option<W>) -> Option<U>
+// - f__N_(x: T, y: S, z: Option<V>, w: W) -> Option<U>
+// - etc.
+// The resulting functions return Some only if all arguments are 'Some'.
+#[macro_export]
+macro_rules! some_function4 {
+    ($func_name:ident, $arg_type0:ty, $arg_type1:ty, $arg_type2: ty, $arg_type3: ty, $ret_type:ty) => {
+        ::paste::paste! {
+            pub fn [<$func_name ___N>]( arg0: $arg_type0, arg1: $arg_type1, arg2: $arg_type2, arg3: Option<$arg_type3> ) -> Option<$ret_type> {
```

Comment:
You could generate these permutations too instead of enumerating them by hand.

Reply:
I am not sure the code will be more readable, but if I need 5 arguments I will consider.